velox/exec/tests/HashJoinTest.cpp

```diff
@@ -5177,6 +5177,80 @@ TEST_F(HashJoinTest, dynamicFiltersAppliedToPreloadedSplits) {
       .run();
 }
 
+TEST_F(HashJoinTest, dynamicFiltersPushDownThroughAgg) {
+  const int32_t numRowsProbe = 300;
+  const int32_t numRowsBuild = 100;
+
+  // Create probe data
+  auto probeVectors = makeBatches(1, [&](int32_t /*unused*/) {
+    return makeRowVector({
+        makeFlatVector<int32_t>(
+            numRowsProbe, [&](auto row) { return row - 10; }),
+        makeFlatVector<int64_t>(numRowsProbe, folly::identity),
+    });
+  });
+  std::shared_ptr<TempFilePath> probeFile = TempFilePath::create();
+  writeToFile(probeFile->getPath(), probeVectors);
+
+  // Create build data
+  auto buildVectors = makeBatches(1, [&](int32_t /*unused*/) {
+    return makeRowVector(
+        {"u0", "u1"},
+        {
+            makeFlatVector<int32_t>(
+                numRowsBuild,
+                [&](auto row) { return 35 + 2 * (row + numRowsBuild / 5); }),
+            makeFlatVector<int64_t>(numRowsBuild, folly::identity),
+        });
+  });
+
+  createDuckDbTable("t", probeVectors);
+  createDuckDbTable("u", buildVectors);
+
+  auto probeType = ROW({"c0", "c1"}, {INTEGER(), BIGINT()});
+  auto planNodeIdGenerator = std::make_shared<core::PlanNodeIdGenerator>();
+  auto buildSide =
+      PlanBuilder(planNodeIdGenerator).values(buildVectors).planNode();
+
+  // Inner join.
+  core::PlanNodeId probeScanId;
+  core::PlanNodeId joinId;
+  auto op = PlanBuilder(planNodeIdGenerator, pool_.get())
+                .tableScan(probeType)
+                .capturePlanNodeId(probeScanId)
+                .partialAggregation({"c0", "c1"}, {"sum(c1)"})
+                .hashJoin(
+                    {"c0", "c1"},
+                    {"u0", "u1"},
+                    buildSide,
+                    "",
+                    {"c0", "c1", "a0"},
+                    core::JoinType::kInner)
+                .capturePlanNodeId(joinId)
+                .planNode();
+
+  SplitInput splitInput = {
+      {probeScanId, {Split(makeHiveConnectorSplit(probeFile->getPath()))}}};
+  HashJoinBuilder(*pool_, duckDbQueryRunner_, driverExecutor_.get())
+      .planNode(std::move(op))
+      .inputSplits(splitInput)
+      .injectSpill(false)
+      .checkSpillStats(false)
+      .referenceQuery(
+          "SELECT c0, c1, sum(c1) FROM t, u WHERE c0 = u0 and c1 = u1 group by c0, c1")
+      .verifier([&](const std::shared_ptr<Task>& task, bool hasSpill) {
+        SCOPED_TRACE(fmt::format("hasSpill:{}", hasSpill));
+        auto planStats = toPlanStats(task->taskStats());
+        auto dynamicFilterStats = planStats.at(probeScanId).dynamicFilterStats;
+        ASSERT_EQ(2, getFiltersProduced(task, /* probeIndex */ 2).sum);
```

Comment:
It might be more readable to use plan node IDs instead of hard-coded indices.

Reply:
Using plan node IDs as inputs may not be precise, as plan node IDs do not necessarily correspond one-to-one with operator indices.

Reply:
Thanks for the detailed explanation. I have added `getOperatorIndex()` to help get the operator index from the plan node id.
```
  // Get the operator index from the plan node id. Only used in the probe-side
  // pipeline. The plan node id starts from "1" and the operator index starts
  // from 0. Plan node IDs map to operators 1:1.
  static int32_t getOperatorIndex(const core::PlanNodeId& planNodeId) {
    return folly::to<int32_t>(planNodeId) - 1;
  }
```